test/extensions/transport_sockets/tls/ecdsa_private_key_method_provider.cc

```diff
@@ -0,0 +1,180 @@
+#include "test/extensions/transport_sockets/tls/ecdsa_private_key_method_provider.h"
+
+#include <memory>
+
+#include "envoy/api/api.h"
+
+#include "openssl/ssl.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace PrivateKeyMethodProvider {
+
+int EcdsaPrivateKeyMethodProvider::ssl_ecdsa_connection_index = -1;
+
+void EcdsaPrivateKeyConnection::delayed_op() {
+  const std::chrono::milliseconds timeout_0ms{0};
+
+  timer_ = dispatcher_.createTimer([this]() -> void {
+    finished_ = true;
+    this->cb_.complete();
+    return;
+  });
+  timer_->enableTimer(timeout_0ms);
+}
+
+static ssl_private_key_result_t privateKeySign(SSL* ssl, uint8_t* out, size_t* out_len,
+                                               size_t max_out, uint16_t signature_algorithm,
+                                               const uint8_t* in, size_t in_len) {
+  (void)out;
+  (void)out_len;
+  (void)signature_algorithm;
+  unsigned char hash[EVP_MAX_MD_SIZE];
+  unsigned int hash_len;
+  EC_KEY* ec_key;
+  const EVP_MD* md;
+  bssl::ScopedEVP_MD_CTX ctx;
+  EcdsaPrivateKeyConnection* ops = static_cast<EcdsaPrivateKeyConnection*>(
+      SSL_get_ex_data(ssl, EcdsaPrivateKeyMethodProvider::ssl_ecdsa_connection_index));
+
+  if (!ops) {
+    return ssl_private_key_failure;
+  }
+
+  ec_key = ops->getPrivateKey();
```

Comment:
there is `bssl::UniquePtr<EC_KEY>`, which will save you from having to remember to `EC_KEY_free` at each return point.

Reply:
Ok, that's a good tip.